### PR TITLE
Add number of items based record multiline formatter

### PR DESCRIPTION
**Pull request by Fizzixnerd**
See https://github.com/fsprojects/fantomas/issues/1143.

I originally wrote the bulk of the code back when I was still calling it "LogicalSize".  I have gone through the diffs looking to make sure I removed all references to that, but just be aware if I missed one that I'm not trying to sneak it in!

Currently the tests assume the current behaviour discussed in https://github.com/fsprojects/fantomas/issues/1167.  If that changes, then one of the tests here will need to change as well.  I have affixed a `FIXME` to the test in question that can be removed when that issue is closed, but please let me know if you want it done a different way.

### CODE REVIEW

**Comment by nojaf** [comment]
@Fizzixnerd I'll review this thoroughly another time but I already want to point out that records have a broader scope.
At first glance, I didn't see any tests for signature files and nothing related to object expressions.
Please consider test-driven development for this project, it can help in not introducing changes without test coverage.

**Comment by Fizzixnerd** [comment]
> @Fizzixnerd I'll review this thoroughly another time but I already want to point out that records have a broader scope.
> At first glance, I didn't see any tests for signature files and nothing related to object expressions.
> Please consider test-driven development for this project, it can help in not introducing changes without test coverage.

Thanks @nojaf.

Yes, I am sort of winging it with what needs to be changed.  I don't think I've ever explicitly used signature files or object expressions before, so it didn't occur to me they would be affected.  I will write tests first for this project from now on, but if I'm not aware of everything that could be affected I'm afraid it won't be enough.  Is there an issue/PR in the past that touched all aspects of records that I could use as a reference?  I obviously don't want to miss anything else.

**Comment by nojaf** [comment]
> I don't think I've ever explicitly used signature files or object expressions

Yeah, I know I'm asking more than you might need yourself but it is to remain consistent.
I would take a look at [MultilineBlockBracketsOnSameColumnRecordTests.fs](https://github.com/fsprojects/fantomas/blob/e2a6c0a8fc572cfabba602bfd7aedb20ddb553e9/src/Fantomas.Tests/MultilineBlockBracketsOnSameColumnRecordTests.fs).

You can write tests for signature files by using `formatSourceString true`.

File: src/Fantomas.Tests/NumberOfItemsListOrArrayTests.fs [hunk 1]
```diff
@@ -1,10 +1,14 @@
-module Fantomas.Tests.NumberOfItemsTests
+module Fantomas.Tests.NumberOfItemsListOrArrayTests
 
 open NUnit.Framework
 open FsUnit
 open Fantomas.Tests.TestHelper
 open Fantomas.FormatConfig
 
+// ****************
```
Comment (nojaf) [inline src/Fantomas.Tests/NumberOfItemsListOrArrayTests.fs:8]: This comment can be removed as you split the file into two modules.

File: src/Fantomas.Tests/NumberOfItemsRecordTests.fs [hunk 2]
```diff
@@ -0,0 +1,982 @@
...
+let g (x: {| x: AReallyLongTypeThatIsMuchLongerThan40Characters |}) = x
+
+type A =
+    {| x: int
+       y: obj |}
+
+type B = {| x: AReallyLongTypeThatIsMuchLongerThan40Characters |}
+"""
+
+// FIXME: Change with https://github.com/fsprojects/fantomas/issues/1167
```
Comment (nojaf) [inline src/Fantomas.Tests/NumberOfItemsRecordTests.fs:957]: Please also ignore this test for now. The result now is indeed not correct as stated in #1167.

**Review by nojaf — APPROVED** [review]

**Comment by nojaf** [comment]
Great PR @Fizzixnerd! And thanks for reporting the inconsistencies you found along the way.

**Comment by Fizzixnerd** [comment]
No problem, thanks for fixing those nitpicks.

**Pull request merged**